core/src/test/scala/unit/kafka/log/LogTest.scala

```diff
@@ -956,6 +963,176 @@ class LogTest extends JUnitSuite {
 
     log.delete()
     assertEquals("The number of segments should be 0", 0, log.numberOfSegments)
-    assertEquals("The number of deleted segments shoud be zero.", 0, log.deleteOldSegments(_ => true))
+    assertEquals("The number of deleted segments shoud be zero.", 0, log.deleteOldSegments())
+  }
+
+  @Test
+  def shouldHaveSizeBasedSegmentsReadyToBeDeleted(): Unit = {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 10)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    assertTrue("There should be segments ready to be deleted", log.shouldDeleteSegments)
+  }
+
+  @Test
+  def shouldDeleteSizeBasedSegments(): Unit = {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 10)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    log.deleteOldSegments
+    assertEquals("should have 2 segments", 2,log.numberOfSegments)
+  }
+
+  @Test
+  def shouldNotDeleteSizeBasedSegmentsWhenUnderRetentionSize(): Unit = {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 15)
+
+    // append some messages to create some segments
+    for (i <- 0 until 15)
+      log.append(set)
+
+    log.deleteOldSegments
+    assertEquals("should have 3 segments", 3,log.numberOfSegments)
+  }
+
+  @Test
+  def shouldNotHaveSizeBasedSegmentsReadyToBeDeleted(): Unit = {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (_, log: Log) = createLog(set.sizeInBytes, retentionBytes = set.sizeInBytes * 10)
+
+    // append some messages to create some segments
+    for (i <- 0 until 8)
+      log.append(set)
+
+    assertFalse("There should be no segments ready to be deleted", log.shouldDeleteSegments)
+  }
+
+
+  @Test
+  def shouldHaveTimeBasedSegmentsReadyToBeDeleted(): Unit = {
+    val set = TestUtils.singleMessageSet("test".getBytes)
+    val (myTime: MockTime, log: Log) = createLog(set.sizeInBytes, retentionMs = 10)
```

Comment:
You are actually pattern matching in this (and other similar) code, not enforcing a type. Spot the difference:

```scala
scala> val (s, f): (Int, Int) = (2.0: Any, 4.0: Any)
<console>:11: error: type mismatch;
 found   : Any
 required: Int
       val (s, f): (Int, Int) = (2.0: Any, 4.0: Any)
                                    ^
<console>:11: error: type mismatch;
 found   : Any
 required: Int
       val (s, f): (Int, Int) = (2.0: Any, 4.0: Any)
                                              ^

scala> val (s: Int, f: Int) = (2.0: Any, 4.0: Any)
scala.MatchError: (2.0,4.0) (of class scala.Tuple2)
  ... 32 elided
```

If the type is obvious from the variable name, you're better off just not writing the type. The compiler will still enforce that the type is used in the right way.